pinot-core/src/main/java/org/apache/pinot/core/indexsegment/immutable/ImmutableSegmentImpl.java

```diff
@@ -91,7 +94,10 @@ public SegmentMetadataImpl getSegmentMetadata() {
 
   @Override
   public DataSource getDataSource(String column) {
-    return new ImmutableDataSource(_segmentMetadata.getColumnMetadataFor(column), _indexContainerMap.get(column));
+    ColumnMetadata columnMetadata = _segmentMetadata.getColumnMetadataFor(column);
+    Preconditions.checkNotNull(columnMetadata,
```

Comment:
If a bad column is queried, this check will end up sending a 5xx to the user, right?  Instead of pre-condition check, we should add specific exception that we can look for and return the right error to the user.

Reply:
As mentioned in the issue #5899 , currently a NullPointerException is being thrown back (without any helpful information to debug). Precondition is just a way to improve the debuggability/readability. 

Lemme know if you have a good suggestion for this.